In [1]:
# Importing libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Defining column names

col_names = ['Elevation',
'Aspect',
'Slope',
'Horizontal_Distance_To_Hydrology',
'Vertical_Distance_To_Hydrology',
'Horizontal_Distance_To_Roadways',
'Hillshade_9am',
'Hillshade_Noon',
'Hillshade_3pm',
'Horizontal_Distance_To_Fire_Points',
'Wilderness_Area_1', 'Wilderness_Area_2', 'Wilderness_Area_3', 'Wilderness_Area_4',
'Soil_Type_1', 'Soil_Type_2', 'Soil_Type_3', 'Soil_Type_4', 'Soil_Type_5', 'Soil_Type_6', 'Soil_Type_7', 'Soil_Type_8', 'Soil_Type_9', 'Soil_Type_10', 'Soil_Type_11', 'Soil_Type_12', 'Soil_Type_13', 'Soil_Type_14', 'Soil_Type_15', 'Soil_Type_16', 'Soil_Type_17', 'Soil_Type_18', 'Soil_Type_19', 'Soil_Type_20', 'Soil_Type_21', 'Soil_Type_22', 'Soil_Type_23', 'Soil_Type_24', 'Soil_Type_25', 'Soil_Type_26', 'Soil_Type_27', 'Soil_Type_28', 'Soil_Type_29', 'Soil_Type_30', 'Soil_Type_31', 'Soil_Type_32', 'Soil_Type_33', 'Soil_Type_34', 'Soil_Type_35', 'Soil_Type_36', 'Soil_Type_37', 'Soil_Type_38', 'Soil_Type_39', 'Soil_Type_40', 
'Cover_Type']

In [3]:
data_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/covtype/covtype.data.gz"

dataset = pd.read_csv(filepath_or_buffer = data_url, compression = 'gzip', 
                      header = None, names = col_names)

In [4]:
dataset.head()

Elevation  Aspect  Slope  Horizontal_Distance_To_Hydrology  \
0       2596      51      3                               258   
1       2590      56      2                               212   
2       2804     139      9                               268   
3       2785     155     18                               242   
4       2595      45      2                               153   

   Vertical_Distance_To_Hydrology  Horizontal_Distance_To_Roadways  \
0                               0                              510   
1                              -6                              390   
2                              65                             3180   
3                             118                             3090   
4                              -1                              391   

   Hillshade_9am  Hillshade_Noon  Hillshade_3pm  \
0            221             232            148   
1            220             235            151   
2            234             238            135   
3            238             238            122   
4            220             234            150   

   Horizontal_Distance_To_Fire_Points  ...  Soil_Type_32  Soil_Type_33  \
0                                6279  ...             0             0   
1                                6225  ...             0             0   
2                                6121  ...             0             0   
3                                6211  ...             0             0   
4                                6172  ...             0             0   

   Soil_Type_34  Soil_Type_35  Soil_Type_36  Soil_Type_37  Soil_Type_38  \
0             0             0             0             0             0   
1             0             0             0             0             0   
2             0             0             0             0             0   
3             0             0             0             0             0   
4             0             0             0             0             0   

   Soil_Type_39  Soil_Type_40  Cover_Type  
0             0             0           5  
1             0             0           5  
2             0             0           2  
3             0             0           2  
4             0             0           5  

[5 rows x 55 columns]

In [5]:
dataset.shape

(581012, 55)

In [6]:
# Summary of dataset

dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 581012 entries, 0 to 581011
Data columns (total 55 columns):
Elevation                             581012 non-null int64
Aspect                                581012 non-null int64
Slope                                 581012 non-null int64
Horizontal_Distance_To_Hydrology      581012 non-null int64
Vertical_Distance_To_Hydrology        581012 non-null int64
Horizontal_Distance_To_Roadways       581012 non-null int64
Hillshade_9am                         581012 non-null int64
Hillshade_Noon                        581012 non-null int64
Hillshade_3pm                         581012 non-null int64
Horizontal_Distance_To_Fire_Points    581012 non-null int64
Wilderness_Area_1                     581012 non-null int64
Wilderness_Area_2                     581012 non-null int64
Wilderness_Area_3                     581012 non-null int64
Wilderness_Area_4                     581012 non-null int64
Soil_Type_1                           581012 non-

###### Summary states that there are 
- 581012 rows
- 55 columns or features
- No null values
- Dtypes of all the entries is int64

In [7]:
# Defining independent and dependent variable

X = dataset.drop(columns = ['Cover_Type'], axis = 1)
y = dataset['Cover_Type']

In [8]:
# Feature Scaling

from sklearn.preprocessing import StandardScaler

sc_X = StandardScaler()
X[:] = sc_X.fit_transform(X[:])

In [9]:
# Splitting data into training and testing sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.25, random_state = 1)

In [10]:
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import confusion_matrix, accuracy_score, classification_report

In [11]:
%%time
# Creating and fitting the model
model_rf = RandomForestClassifier(n_estimators = 50, oob_score = True,
                              n_jobs = -1)
model_rf.fit(X_train, y_train)

# Predicting test set results
y_pred = model_rf.predict(X_test)

# Predicting training set results
x_pred = model_rf.predict(X_train)

# Out Of the Box score of this model
oob_score = model_rf.oob_score_

# Creating confusion matrix
cm = confusion_matrix(y_test, y_pred)

# Training set accuracy
accuracy_x = accuracy_score(y_train, x_pred)

# Test set accuracy
accuracy_y = accuracy_score(y_test, y_pred)

# Print accuracies and OOB score
print(f'Training Set Accuracy = {accuracy_x:.3f} \tTest Set Accuracy = {accuracy_y:.3f}')

print(f'\nOOB Score = {oob_score:.3f}')

Training Set Accuracy = 1.000        	Test Set Accuracy = 0.953

OOB Score = 0.948
Wall time: 1min 7s


In [12]:
# Feature Importance
fi = model_rf.feature_importances_

# Convert fi array into dataframe
fi_df = pd.DataFrame(data = fi, 
                     index = X_train.columns, 
                     columns = ['Importance'])

In [13]:
fi_df.sort_values(by = 'Importance', ascending = False)

Importance
Elevation                             0.243212
Horizontal_Distance_To_Roadways       0.118065
Horizontal_Distance_To_Fire_Points    0.109321
Horizontal_Distance_To_Hydrology      0.060857
Vertical_Distance_To_Hydrology        0.057249
Aspect                                0.047634
Hillshade_Noon                        0.043833
Hillshade_9am                         0.041434
Hillshade_3pm                         0.041426
Slope                                 0.032452
Wilderness_Area_4                     0.029842
Soil_Type_22                          0.016695
Soil_Type_4                           0.012430
Wilderness_Area_1                     0.011627
Wilderness_Area_3                     0.011478
Soil_Type_10                          0.011429
Soil_Type_23                          0.010707
Soil_Type_12                          0.010552
Soil_Type_39                          0.009866
Soil_Type_38                          0.009382
Soil_Type_2                           0.009330
Soil_Type_29                          0.005937
Soil_Type_32                          0.005781
Wilderness_Area_2                     0.005596
Soil_Type_40                          0.004482
Soil_Type_33                          0.004369
Soil_Type_13                          0.004107
Soil_Type_24                          0.004079
Soil_Type_31                          0.003843
Soil_Type_30                          0.003024
Soil_Type_11                          0.002580
Soil_Type_3                           0.002449
Soil_Type_6                           0.002160
Soil_Type_20                          0.001937
Soil_Type_17                          0.001607
Soil_Type_35                          0.001489
Soil_Type_19                          0.001011
Soil_Type_1                           0.000841
Soil_Type_16                          0.000818
Soil_Type_21                          0.000771
Soil_Type_5                           0.000726
Soil_Type_27                          0.000659
Soil_Type_37                          0.000584
Soil_Type_34                          0.000541
Soil_Type_14                          0.000425
Soil_Type_26                          0.000423
Soil_Type_18                          0.000300
Soil_Type_25                          0.000252
Soil_Type_28                          0.000152
Soil_Type_9                           0.000116
Soil_Type_36                          0.000069
Soil_Type_8                           0.000034
Soil_Type_7                           0.000011
Soil_Type_15                          0.000004

In [14]:
# Display confusion matrix as a dataframe
cm_df = pd.DataFrame(data = cm, 
             index = ['Actual Covertype 1', 'Actual Covertype 2', 'Actual Covertype 3', 'Actual Covertype 4', 'Actual Covertype 5', 'Actual Covertype 6', 'Actual Covertype 7'], 
             columns = ['Predicted Covertype 1', 'Predicted Covertype 2', 'Predicted Covertype 3', 'Predicted Covertype 4', 'Predicted Covertype 5', 'Predicted Covertype 6', 'Predicted Covertype 7']
            )
cm_df

Predicted Covertype 1  Predicted Covertype 2  \
Actual Covertype 1                  49945                   2840   
Actual Covertype 2                   1696                  68704   
Actual Covertype 3                      1                    145   
Actual Covertype 4                      0                      0   
Actual Covertype 5                     35                    496   
Actual Covertype 6                      6                    121   
Actual Covertype 7                    250                     25   

                    Predicted Covertype 3  Predicted Covertype 4  \
Actual Covertype 1                      1                      0   
Actual Covertype 2                    167                      1   
Actual Covertype 3                   8637                     29   
Actual Covertype 4                     86                    585   
Actual Covertype 5                     27                      0   
Actual Covertype 6                    336                     22   
Actual Covertype 7                      0                      0   

                    Predicted Covertype 5  Predicted Covertype 6  \
Actual Covertype 1                     31                      3   
Actual Covertype 2                     88                    100   
Actual Covertype 3                      8                    189   
Actual Covertype 4                      0                     14   
Actual Covertype 5                   1785                      9   
Actual Covertype 6                      2                   3881   
Actual Covertype 7                      2                      0   

                    Predicted Covertype 7  
Actual Covertype 1                    117  
Actual Covertype 2                     25  
Actual Covertype 3                      0  
Actual Covertype 4                      0  
Actual Covertype 5                      0  
Actual Covertype 6                      0  
Actual Covertype 7                   4844

In [15]:
# Classification Report
class_names = ['Cover Type 1', 'Cover Type 2', 'Cover Type 3', 
               'Cover Type 4', 'Cover Type 5', 'Cover Type 6', 
               'Cover Type 7']

c_report = classification_report(y_true = y_test, y_pred = y_pred, target_names = class_names, output_dict = True)

In [16]:
report_df = pd.DataFrame(data = c_report)
report_df

Cover Type 1  Cover Type 2  Cover Type 3  Cover Type 4  \
f1-score       0.952513      0.960143      0.945847      0.885023   
precision      0.961720      0.949856      0.933326      0.918367   
recall         0.943480      0.970656      0.958708      0.854015   
support    52937.000000  70781.000000   9009.000000    685.000000   

           Cover Type 5  Cover Type 6  Cover Type 7  accuracy      macro avg  \
f1-score       0.836457      0.906352      0.958544  0.952689       0.920697   
precision      0.931628      0.924929      0.971520  0.952689       0.941621   
recall         0.758929      0.888507      0.945909  0.952689       0.902886   
support     2352.000000   4368.000000   5121.000000  0.952689  145253.000000   

            weighted avg  
f1-score        0.952444  
precision       0.952725  
recall          0.952689  
support    145253.000000

In [17]:
%%time

# Calculating cross validation scores

from sklearn.model_selection import cross_val_score

accuracies = cross_val_score(estimator = model_rf, 
                             X = X_train, y = y_train, 
                             cv = 5, 
                             n_jobs = -1)

print("Cross Validation (5 fold)")
print("=" * 50)
print(f'Mean Accuracy = {accuracies.mean() * 100:.3f}%')
print(f'Variance = {accuracies.std() * 100:.3f}%')

Cross Validation (5 fold)
Mean Accuracy = 94.689%
Variance = 0.064%
Wall time: 3min 24s


###### After evaluating model's OOB score, Confusion Matix, Classification Report and Cross Validation Scores, we can conclude that the model is performing well.